ion-c-sys/src/result.rs

```diff
@@ -10,11 +10,66 @@ use std::fmt;
 use std::num::TryFromIntError;
 
 /// IonC Error code and its associated error message.
+///
+/// `position` is populated when errors come from readers. See [`Position`] for
+/// more information.
 #[derive(Copy, Clone, Debug, PartialEq)]
 pub struct IonCError {
     pub code: i32,
     pub message: &'static str,
     pub additional: &'static str,
+    pub position: Option<Position>,
+}
+
+/// Represents a position in a data source. For example, consider a file
+/// containing Ion data that is being parsed using an [`IonCReader`].
+///
+/// `bytes` will always be hydrated while `lines` and `offset` will only be
+/// populated for text readers.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Position {
+    pub bytes: i64,
+    pub line: Option<i32>,
+    pub offset: Option<i32>,
+}
+
+impl Position {
+    pub fn new(bytes: i64, line: i32, offset: i32) -> Position {
+        Position {
+            bytes,
+            line: Some(line),
+            offset: Some(offset),
+        }
+    }
+
+    pub fn bytes(bytes: i64) -> Position {
+        Position {
+            bytes,
+            line: None,
+            offset: None,
+        }
+    }
+}
+
+impl From<(i64, i32, i32)> for Position {
+    fn from(ionc_values: (i64, i32, i32)) -> Self {
+        let bytes = ionc_values.0;
+        let lines = if ionc_values.1 < 0 {
+            None
+        } else {
+            Some(ionc_values.1)
+        };
+        let offset = if ionc_values.2 < 0 {
+            None
+        } else {
+            Some(ionc_values.2)
+        };
+        Self {
+            bytes,
+            line: lines,
+            offset,
+        }
+    }
```

Comment:
I see what is going on here--though I think it better to just put into a named method.  This seems a little clever to have a call-site like `(...).into()` do this conversion via negatives.

Reply:
I'll remove this method and use the constructors directly.